OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,314 @@
+					var tempSwingDirection = Game.CosmeticRandom.Next(2) == 0 ? 1 : -1;
+
+					particleList.Add(
+						new Particle
+						{
+							PosX = Game.CosmeticRandom.Next(Game.Renderer.Resolution.Width),
+							PosY = Game.CosmeticRandom.Next(rangeY),
+							Size = Game.CosmeticRandom.Next(info.PaticleSize[0], info.PaticleSize[1] + 1),
+							DirectionScatterX = info.ScatterDirection[0] + Game.CosmeticRandom.Next(info.ScatterDirection[1] - info.ScatterDirection[0]),
+							Gravity = float2.Lerp(info.Gravity[0], info.Gravity[1], Game.CosmeticRandom.NextFloat()),
+							SwingOffset = float2.Lerp(info.SwingOffset[0], info.SwingOffset[1], Game.CosmeticRandom.NextFloat()),
+							SwingSpeed = float2.Lerp(info.SwingSpeed[0], info.SwingSpeed[1], Game.CosmeticRandom.NextFloat()),
+							SwingDirection = tempSwingDirection,
+							SwingAmplitude = float2.Lerp(info.SwingAmplitude[0], info.SwingAmplitude[1], Game.CosmeticRandom.NextFloat()),
+							Color = tempColor,
+							TailColor = tempColorTail
+						});
+				}
+			}
+		}
+
+		void ParticlesCountLogic(WorldRenderer wr)
+		{
+			// Logic to switch between the states of the particleCountFader
+			if (particleCountFader == ParticleCountFaderType.Hold && particleList.Count < maxParticleCount)
+				particleCountFader = ParticleCountFaderType.FadeIn;
+			else if (particleCountFader == ParticleCountFaderType.FadeIn && particleList.Count >= maxParticleCount)
+				particleCountFader = ParticleCountFaderType.Hold;
+			else if (particleCountFader == ParticleCountFaderType.FadeOut && particleList.Count == 0)
+				particleCountFader = ParticleCountFaderType.Hold;
+
+			// Do the fade functions
+			if (particleCountFader == ParticleCountFaderType.FadeIn)
+				FadeInParticleCount(wr);
+			else if (particleCountFader == ParticleCountFaderType.FadeOut)
+				FadeOutParticleCount(wr);
+		}
+
+		void FadeInParticleCount(WorldRenderer wr)
+		{
+			SpawnParticles(1, 0, 100);
+
+			// Remove Particles, which are getting replaced from the top to the bottom by the "EdgeCheckReplace",
+			// when scrolling down, as long as the FadeIn is not completed,
+			// to avoid having particles at the top and bottom, but not in the middle of the screen.
+			for (var i = 0; i < particleList.Count; i++)
+			{
+				if (particleList[i].PosY < 0)
+					particleList.RemoveAt(i);
+			}
+
+			// Add Particles when the weather is fading in and scrolling up, to fill areas above
+			if (antiScrollPrevTopLeft.Y > wr.Viewport.TopLeft.Y)
+			{
+				// Get delta Y and limit to the max value
+				var tempRangeY = antiScrollPrevTopLeft.Y - wr.Viewport.TopLeft.Y;
+				var tempParticleCount = CalculateParticleCount(Game.Renderer.Resolution.Width, (int)tempRangeY);
+				if (particleList.Count + tempParticleCount > maxParticleCount)
+					tempParticleCount = maxParticleCount - particleList.Count;
+
+				SpawnParticles(tempParticleCount, (int)tempRangeY, 50);
```

Comment:
its a 50:50 spawn chance. Other Values would not be very usefull and like to look ugly. 
It's like scrolling upwards and getting a brick of white snowflakes in the face, or erasing lines in the  same way.